File: java-way/src/main/java/com/crack/snap/make/easy/linkedlist/ReverseLinkedList.java
package com.crack.snap.make.easy.linkedlist;

/**
 * @author Mohan Sharma
 */
public class ReverseLinkedList {

    public ListNode reverseListIterative(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode current = head, previous = null;
        while (current != null) {
            var next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        return previous;
    }

    public ListNode reverseListRecursive(ListNode head) {
        if (head == null) {
            return null;
        }
        return reverseListRecursiveHelper(head, null);
    }

    private ListNode reverseListRecursiveHelper(ListNode current, ListNode prev) {
        if (current == null) {
            return prev;
        }
        var next = current.next;
        current.next = prev;
        return reverseListRecursiveHelper(next, current);
    }

    public static void main(String[] args) {
        var list = new LinkedList();
        list.addLast(1);
        list.addLast(2);
        list.addLast(3);
        list.addLast(4);
        list.addLast(5);
        list.print(list.head);
        var reverseLinkedList = new ReverseLinkedList();
        var head = reverseLinkedList.reverseListRecursive(list.head);
        list.print(head);
    }
}

File: python-way/com/crack/snap/make/easy/linkedlist/reverse_linked_list.py
# @author Mohan Sharma

from typing import Optional

from com.crack.snap.make.easy.linkedlist.list_node import ListNode


def reverseListIterative(head: Optional[ListNode]) -> Optional[ListNode]:
	current_node = head
	previous_node = None
	while current_node:
		next_node = current_node.next
		current_node.next = previous_node
		previous_node = current_node
		current_node = next_node
	return previous_node

def reverseListRecursive(head: Optional[ListNode]) -> Optional[ListNode]:
	def helper(current_node, previous_node):
		if not current_node:
			return previous_node
		next_node = current_node.next
		current_node.next = previous_node
		return helper(next_node, current_node)
	return helper(head, None)

def reverseListRecursiveWithoutHelper(head: Optional[ListNode]) -> Optional[ListNode]:
	if not head or not head.next:
		return head
	reversed_head = reverseListRecursiveWithoutHelper(head.next)
	head.next.next = head
	head.next = None
	return reversed_head
